clusterDeployer.py

```diff
@@ -33,6 +34,104 @@ def match_to_proper_version_format(version_cluster_config: str) -> str:
+    #
+    # We will still write our "/etc/resolv.conf.cda*" files, but
+    # it's up to the user to redirect symlinks accordingly.
+    dont_touch = os.path.exists("/etc/.resolv.conf.cda-dont-touch")
+
+    if cluster_name is None:
+        if dont_touch:
+            return
+        if not os.path.exists(rc_orig):
+            return
+        try:
+            lnk = os.readlink(rc)
+        except Exception:
+            return
+        if not os.path.basename(lnk).startswith("resolv.conf.cda."):
+            # We only restore, if /etc/resolv.conf is a symlink to "resolv.conf-cda.*"
+            # which indicates that the file was created by us.
+            return
+        dst = os.path.basename(rc_orig)
+        if lnk == dst:
+            return
+        logger.info(f"restore /etc/resolv.conf to link to original {rc_orig}")
+        lh.run_or_die(shlex.join(["ln", "-snf", dst, rc]))
+        return
+
+    rc_my = f"/etc/resolv.conf.cda.{cluster_name}"
+
+    logger.info(f"updating /etc/resolv.conf to link to {rc_my} (nameserver: {api_vip})")
+
+    content_rc = b''
+    if not os.path.exists(rc_orig):
+        try:
+            with open(rc, "rb") as f:
+                content_rc = f.read()
+        except IOError:
+            pass
+        # Backup the original resolv.conf file to "rc_orig"
+        if b"Written by cluster-deployment-automation" in content_rc:
+            content_rc = b''
+        elif b"Generated by NetworkManager" in content_rc and os.path.exists("/run/NetworkManager/resolv.conf"):
+            # This file seems to be written by NetworkManager. Symlink there.
+            lh.run_or_die(shlex.join(["ln", "-snf", "/run/NetworkManager/resolv.conf", rc_orig]))
+        else:
+            with common.atomic_write(rc_orig, text=False) as f:
+                f.write(content_rc)
+    else:
+        try:
+            with open(rc_orig, "rb") as f:
+                content_rc = f.read()
+        except IOError:
+            pass
+
+    extra_nameservers = []
+    for line in content_rc.split(b'\n'):
+        line = line.strip()
+        if not line.startswith(b"nameserver "):
+            continue
+        line = line[len(b"nameserver ") :]
+        line = line.strip()
+        if not line:
+            continue
+        n = line.decode('utf-8', errors='ignore')
+        if n not in extra_nameservers and n != api_vip:
+            extra_nameservers.append(n)
+
+    extra_nameservers_s = ""
+    if extra_nameservers:
+        extra_nameservers_s = f"\n\n# nameservers from {rc_orig}"
+        for n in extra_nameservers:
+            extra_nameservers_s += f"\nnameserver {n}"
+
+    with common.atomic_write(rc_my) as file:
+        file.write(
+            f"""# Written by cluster-deployment-automation for "{cluster_name}"
+# Original file at {rc_orig}
+# Set:     sudo ln -snf {os.path.basename(rc_my)} /etc/resolv.conf
+# Restore: sudo ln -snf {os.path.basename(rc_orig)} /etc/resolv.conf
+nameserver {api_vip}{extra_nameservers_s}
```

Comment:
This is problematic when running workers in VMs on remote nodes:

> Redirecting all DNS requests to the cluster is maybe problematic. Maybe it would be better to start a local caching DNS server. In any case, you can opt out from this with `sudo touch /etc/.resolv.conf.cda-dont-touch`. It's also simple to switch back and forth.

Those VMs will try to pull images from `registry.redhat.io` via the api_network (192.168.122.X) reach the provisioner node and get forwarded to the API_VIP.. which can't reach `registry.redhat.io`.

Reply:
and VMs cannot resolve those names? The can resolv public DNS names, registry.redhat.io seems reachable from the "internet", isn't it?

Reply:
Maybe we need to set up split dns instead.

Reply:
I get this in the VMs:
```
Apr 08 15:31:38 ovnscale-worker-1 podman[35808]: Error: initializing source docker://registry.redhat.io/rhai-tech-preview/assisted-installer-agent-rhel8:v1.0.0-295: pinging container registry registry.redhat.io: Get "https://registry.red>
Apr 08 15:31:38 ovnscale-worker-1 systemd[1]: agent.service: Control process exited, code=exited, status=125/n/a
Apr 08 15:31:38 ovnscale-worker-1 systemd[1]: agent.service: Failed with result 'exit-code'.
Apr 08 15:31:38 ovnscale-worker-1 systemd[1]: Failed to start agent.service.
```

Reply:
VMs have:
```
# Generated by NetworkManager
nameserver 192.168.122.1
```
which is the IP of the provisioner

Reply:
You are right.

Let me rework this to use dnsmasq...

Closing this one. Will follow up.

Reply:
I thought that I sufficiently tested this approach, and that it worked for me.

That is, I could use the nameserver from `"api.{cluster_name}.redhat.com"` to resolve names. I probably didn't test, that that name server couldn't resolve the internet (didn't cross my mind)...